File: cpu/mc1322x/tools/rftestrx2pcap.pl
#!/usr/bin/perl -w

use Device::SerialPort;
use Term::ReadKey;
use Getopt::Long;
use Time::HiRes qw(usleep gettimeofday);

use strict;

my $filename = '';
my $second = '';
my $term = '/dev/ttyUSB0';
my $baud = '115200';
my $verbose;

GetOptions (
	    'terminal=s' => \$term, 
	    'baud=s' => \$baud,
    );

$| = 1;

# TODO: add help argument
#    print "Example usage: rftestrx2pcap.pl -t /dev/ttyS0 -b 9600\n";
#    exit;

my $ob = Device::SerialPort->new ($term) or die "Can't start $term\n";
    # next test will die at runtime unless $ob

$ob->baudrate($baud);
$ob->parity('none');
$ob->databits(8);
$ob->stopbits(1);
$ob->read_const_time(1000); # 1 second per unfulfilled "read" call

my $str = '';
my ($sec, $usec, $len);
my @frame;

my $magic = 0xa1b2c3d4;
my $major = 2;
my $minor = 4;
my $zone = 0;
my $sig = 0;
my $snaplen = 0xffff;
my $network = 230; # 802.15.4 no FCS

my $newpacket = 0;
my $len = 0;

print pack('LSSLLLL',($magic,$major,$minor,$zone,$sig,$snaplen,$network));

while(1) {
    my ($count, $c) = $ob->read(1);

    if (defined($count) && ($count != 0)) {
	$str .= $c;
	# match if ends in \n or \r and process line
	if(($str =~ /\n$/) ||
	   ($str =~ /\r$/)) {
	    if($str =~ /^rftest-rx --- len 0x(\w\w)/) {
		#new packet
		($sec, $usec) = gettimeofday;
		$len = hex($1);
		print STDERR "rftestline: $sec $usec $len $str";		
		$newpacket = 1;
	    } elsif($str =~ /^\w+/) {		
		# dataline, write out pcap entry
		$str =~ /(.+)/;
		chomp $str;
		my @data = split(' ',$1);

		# do header if this is a new packet
		if($newpacket == 1) {
		    $newpacket = 0;
		    print pack('LLLL',($sec,$usec,$len,$len));
		    print STDERR "new packet: $sec $usec $len " . ($len) . "\n\r";
		}

		# packet payload
		print STDERR "dataline: ";		
		print STDERR $str . "\n\r";

		foreach my $data (@data) {
		    print pack ('C',hex($data));
		}		
	    }
	    $str = '';
	}
    }
}

$ob -> close or die "Close failed: $!\n";
ReadMode 0;
undef $ob;  # closes port AND frees memory in perl
exit;

File: cpu/mc1322x/tools/rftestrx2pcap.py
#!/usr/bin/python
# (C) 2012, Mariano Alvira <mar@devl.org>

import sys,os,time
from struct import *
import re
import serial

if len(sys.argv) < 3:
	sys.stderr.write( "Usage: %s tty channel [outfile]\n" %(sys.argv[0]))
        sys.stderr.write( "       channel = 11-26\n")
	sys.exit(2)

# change the channel
# rftest-rx increments it's channel everytime you send a character and returns "channel: num"
# send a character until we get to the right channel

try:
	serport = serial.Serial(sys.argv[1], 115200, timeout=1)
except IOError:
	print "error opening port"
	sys.exit(2)

chan = ''
while chan != int(sys.argv[2]) - 11:
	serport.write(' ')
	chanstr = ''
	while 1:
		chanstr += serport.read(1)
		m = re.match(".*channel: (\w+)\s+", chanstr)
		if m is not None:
			chan = int(m.group(1))
			break

try:
	sys.stderr.write('writing to file %s \n' % (sys.argv[3]))
	outfile = open(sys.argv[3], 'w+b')
except IndexError:
	outfile = sys.stdout

sys.stderr.write("RX: 0\r")

### PCAP setup
MAGIC = 0xa1b2c3d4;
MAJOR = 2;
MINOR = 4;
ZONE = 0;
SIG = 0;
SNAPLEN = 0xffff;
NETWORK = 230; # 802.15.4 no FCS

# output overall PCAP header
outfile.write(pack('<LHHLLLL', MAGIC, MAJOR, MINOR, ZONE, SIG, SNAPLEN, NETWORK))

count = 0
fileempty = 1
newpacket = 0

try:
	while 1:
		line = serport.readline().rstrip()

		m_rftestline = re.match(".*rftest-rx --- len 0x(\w\w).*", line)

		if m_rftestline is not None:
			newpacket = 1
			t = time.time()
			sec = int(t)
			usec = (t - sec) * 100000
			length = int(m_rftestline.group(1), 16)
#			sys.stderr.write(line + "\n")
#			sys.stderr.write("rftestline: %d %d %d\n" % (sec, usec, length))
			continue
			
			# if this is a new packet, add a packet header
		if newpacket == 1:
			newpacket = 0
			outfile.write(pack('<LLLL',sec,usec,length,length))

			count += 1
			sys.stderr.write("RX: %d\r" % count)			

			# clear file empty flag
			if fileempty:
				fileempty = 0
		if fileempty == 0 :
			# write payload
			for d in line.split(' '):
				# do a match because their might be a \r floating around
				m = re.match('.*(\w\w).*', d)
				if m is not None:
					outfile.write(pack('<B', int(m.group(1),16)))


#             cn.recv_block()
#             if cn.data != None:

# #                sys.stderr.write(cn.data)
# #                sys.stderr.write(cn.data)
# #                sys.stderr.write(str(time.time()))

#                 t = time.time()
#                 sec = int(t)
#                 usec = (t - sec) * 100000

# #                sys.stderr.write("new packet: %d %d %d" % (sec, usec, len(cn.data)))

#                 count += 1
#                 sys.stderr.write("\x1b[G") # Move the cursor up one
#                 sys.stderr.write("RX: %d" % count)

#                 sys.stdout.write(pack('<LLLL',sec,usec,len(cn.data),len(cn.data)))
#                 sys.stdout.write(cn.data)

except KeyboardInterrupt:
#		cn.close()
		sys.exit(2)


